samples/nrf9160/http_application_update/src/main.c

```diff
@@ -0,0 +1,231 @@
+
+	switch (event) {
+	case DOWNLOAD_CLIENT_EVT_DOWNLOAD_FRAG: {
+		struct flash_pages_info info;
+		static u32_t address = CONFIG_FLASH_WRITE_OFFSET;
+
+		err = flash_get_page_info_by_offs(flash_dev, address, &info);
+		if (err != 0) {
+			__ASSERT(false,
+				"flash_get_page_info_by_offs returned error %d\n",
+				err);
+		}
+		flash_write_protection_set(flash_dev, false);
+		if (!is_flash_page_erased[info.index]) {
+			err = flash_erase(flash_dev, address, info.size);
+			if (err != 0) {
+				__ASSERT(false,
+					"flash_erase returned error %d at address %08x\n",
+					err, address);
+			}
+			is_flash_page_erased[info.index] = true;
+		}
+		err = flash_write(flash_dev, address, dfu->fragment,
+							dfu->fragment_size);
+		flash_write_protection_set(flash_dev, true);
+		if (err != 0) {
+			__ASSERT(false,
+				"Flash write error %d at address %08x\n",
+				err, address);
+		}
+		address += dfu->fragment_size;
+		break;
+	}
+
+	case DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE:
+		LOG_INF("DOWNLOAD_CLIENT_EVT_DOWNLOAD_DONE");
+		break;
+
+	case DOWNLOAD_CLIENT_EVT_ERROR: {
+		LOG_ERR("Download client error");
+		download_client_disconnect(dfu);
+		__ASSERT(false, "Something went wrong, please restart "
+			"the application\n");
+		break;
+	}
+	default:
+		break;
+	}
+	return 0;
+}
+
+
+/**@brief Turn on LED0 and LED1 if CONFIG_APPLICATION_VERSION
+ * is 2 and LED0 otherwise.
+ */
+static void led_app_version(void)
+{
+	struct device *dev;
+
+	dev = device_get_binding(LED_PORT);
+	__ASSERT(dev != 0, "Nordic nRF GPIO driver was not found!\n");
+
+	gpio_pin_configure(dev, LED0_GPIO_PIN, GPIO_DIR_OUT);
+	gpio_pin_write(dev, LED0_GPIO_PIN, 1);
+
+#if CONFIG_APPLICATION_VERSION == 2
+	gpio_pin_configure(dev, LED1_GPIO_PIN, GPIO_DIR_OUT);
+	gpio_pin_write(dev, LED1_GPIO_PIN, 1);
+#endif
+}
+
+static void dfu_worker_thread(void *p1, void *p2, void *p3)
+{
+	app_dfu_init();
+	app_dfu_transfer_start();
+
+	while (dfu.status != DOWNLOAD_CLIENT_STATUS_DOWNLOAD_COMPLETE) {
+		download_client_process(&dfu);
+	}
+
+	download_client_disconnect(&dfu);
+	LOG_INF("Download complete");
+}
+
+void dfu_button_pressed(struct device *gpiob, struct gpio_callback *cb,
+		    u32_t pins)
+{
+	LOG_INF("dfu_button_pressed");
+	start_dfu = true;
+}
+
+static void dfu_button_init(void)
+{
+	int err;
+	struct device *gpiob;
+
+	gpiob = device_get_binding(SW0_GPIO_CONTROLLER);
+	__ASSERT(gpiob != 0, "Nordic nRF GPIO driver was not found!\n");
+	err = gpio_pin_configure(gpiob, SW0_GPIO_PIN,
+			GPIO_DIR_IN		|
+			GPIO_INT		|
+			GPIO_PUD_PULL_UP|
+			GPIO_INT_EDGE	|
+			GPIO_INT_ACTIVE_LOW);
+	if (err == 0) {
+		gpio_init_callback(&gpio_cb, dfu_button_pressed,
+			BIT(SW0_GPIO_PIN));
+		err = gpio_add_callback(gpiob, &gpio_cb);
+	}
+	if (err == 0) {
+		err = gpio_pin_enable_callback(gpiob, SW0_GPIO_PIN);
+	}
+	__ASSERT(err == 0, "Unable to configure SW0_GPIO_PIN!\n");
+}
+
+static void start_dfu_thread(void)
+{
+	start_dfu = false;
+	k_thread_create(&dfu_thread, dfu_thread_stack,
```

Comment:
When pressing the button, you should start executing the thread. I think you do not want to create multiple threads here. You are using the same event handler for the DFU client.

Reply:
>When pressing the button, you should start executing the thread.  I think you do not want to create multiple threads here.

I'm not really disagreeing with you on that you could create the thread then `suspend` it and start it on a button press with `resume` or use multi-processing primitives such as semaphores, queues, mutexes etc. I've commented about it here https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/587#discussion_r269963543, but for this sample I don't think this is relevant and with regards to zephyr docs  this should work as you initialize and schedule it. Also this will only happen once before it's reset to boot the downloaded image.

>This routine initializes a thread, then schedules it for execution. https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html?highlight=k_thread_create#_CPPv315k_thread_createP8k_threadP16k_thread_stack_t6size_t16k_thread_entry_tPvPvPvi5u32_t5s32_t

I do see that it may be a bit confusing that a thread is created as you could have done it all in main but  I think the idea about the sample is to show case downloading an firmware image while doing something else? It's currently doing nothing but customers could extend it and remove the button press DFU part with some logic. 

I don't understand the comment 

>You are using the same event handler for the DFU client.

Maybe I'm not looking at the same files maybe the dfu_client.c has also defined `dfu_thread`? I'm not sure how zephyr handles multiple `K_THREAD` defines if it separates the handler based on source file or if it's global. Since it's defined as static `static struct k_thread dfu_thread` I think it should be fine at least the compiler should have complained about it IMO, if not Zephyr should have a build time check that at least warns if not faults when hooking more threads on to the same handle.

Reply:
I have removed threading now.